### PR TITLE
Drop support for Kubernetes < 1.20

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area open-source
/kind cleanup
/platform aws

**What this PR does / why we need it**:
Drop support for kubernetes < 1.20 for GCP extension provider

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/6911

**Special notes for your reviewer**:

**Release note**:

```breaking operator
provider-aws no longer supports Shoots with Кubernetes version < 1.20.
```

### CODE REVIEW

File: pkg/admission/validator/shoot.go [hunk 1]
```diff
@@ -22,10 +22,6 @@ import (
 
 	api "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws"
 	awsvalidation "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/validation"
-	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
```
Comment (ialidzhikov) [inline pkg/admission/validator/shoot.go:25]: Nit: Can you keep the `github.com/gardener/gardener-extension-provider-aws` imports in a separate import group as before?

File: pkg/admission/validator/shoot.go [hunk 2]
```diff
@@ -127,11 +123,6 @@ func (s *shoot) validateShoot(_ context.Context, shoot *core.Shoot) error {
 		return err
 	}
 
-	csiMigrationVersion, err := semver.NewVersion(aws.GetCSIMigrationKubernetesVersion(&extensionscontroller.Cluster{Shoot: shootV1beta1}))
```
Comment (ialidzhikov) [inline pkg/admission/validator/shoot.go:130]: provider-aws is supporting the `alpha.csimigration.shoot.extensions.gardener.cloud/kubernetes-version` annotation for delaying the CSI migration. Before cleaning up the CSI related handling did we checked whether this feature is used on Shoots on our landscapes? You could add at least a release not of type breaking that support for `alpha.csimigration.shoot.extensions.gardener.cloud/kubernetes-version` is being removed and provider-aws will assume that the CSI migration completed in 1.18.

File: pkg/controller/controlplane/valuesprovider.go [hunk 3]
```diff
@@ -349,16 +348,11 @@ func (vp *valuesProvider) GetControlPlaneShootChartValues(
...
 ) (map[string]interface{}, error) {
-	csiEnabled, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, ">=", aws.GetCSIMigrationKubernetesVersion(cluster))
-	if err != nil {
-		return nil, err
-	}
-
 	return map[string]interface{}{
 		"volumesnapshots": map[string]interface{}{
-			"enabled": csiEnabled,
+			"enabled": true,
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider.go:355]: Similar to https://github.com/gardener/gardener-extension-provider-gcp/pull/533#discussion_r1050861636

File: pkg/controller/worker/machines_test.go [hunk 4]
```diff
@@ -27,6 +27,7 @@ import (
 	apiv1alpha1 "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/v1alpha1"
 	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
 	. "github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker"
+	"github.com/gardener/gardener/pkg/utils/test/matchers"
```
Comment (ialidzhikov) [inline pkg/controller/worker/machines_test.go:30]: Nit: You could move the the 3rd import group.

File: pkg/controller/worker/machines_test.go [hunk 5]
```diff
@@ -761,7 +684,8 @@ var _ = Describe("Machines", func() {
 
 					result, err := workerDelegate.GenerateMachineDeployments(ctx)
 					Expect(err).NotTo(HaveOccurred())
-					Expect(result).To(Equal(machineDeployments))
+					_ = machineDeployments
```
Comment (ialidzhikov) [inline pkg/controller/worker/machines_test.go:687]: I assume this is a leftover from testing.

```suggestion
```

File: pkg/controller/worker/machines_test.go [hunk 6]
```diff
@@ -761,7 +684,8 @@ var _ = Describe("Machines", func() {
...
+					Expect(result).To(matchers.DeepEqual(machineDeployments))
```
Comment (ialidzhikov) [inline pkg/controller/worker/machines_test.go:688]: For consistency reasons you could use `matchers.DeepEqual` for the assertion in https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/pkg/controller/worker/machines_test.go#L639

File: pkg/webhook/controlplane/ensurer.go [hunk 7]
```diff
@@ -346,93 +299,42 @@ var (
...
 
-func ensureKubeAPIServerVolumeMounts(c *corev1.Container, csiEnabled, csiMigrationComplete bool) {
-	if csiEnabled && csiMigrationComplete {
-		c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount.Name)
-		return
-	}
-
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount)
+func ensureKubeAPIServerVolumeMounts(c *corev1.Container) {
+	c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount.Name)
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer.go:308]: Similar to https://github.com/gardener/gardener-extension-provider-gcp/pull/533#discussion_r1050822910

File: pkg/webhook/controlplane/ensurer.go [hunk 8]
```diff
@@ -23,10 +23,8 @@ import (
 
 	"github.com/Masterminds/semver"
 	"github.com/coreos/go-systemd/v22/unit"
-	"github.com/gardener/gardener/extensions/pkg/controller/csimigration"
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer.go:26]: https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/pkg/webhook/controlplane/ensurer.go#L153-L161 could be also cleaned up.

File: pkg/webhook/controlplane/ensurer_test.go [hunk 9]
```diff
@@ -202,38 +118,18 @@ var _ = Describe("Ensurer", func() {
...
-			client.EXPECT().Get(ctx, secretKey, &corev1.Secret{}).DoAndReturn(clientGet(secret))
-			client.EXPECT().Get(ctx, cmKey, &corev1.ConfigMap{}).DoAndReturn(clientGet(cm))
-
-			err := ensurer.EnsureKubeAPIServerDeployment(ctx, eContextK8s117, dep, nil)
-			Expect(err).To(Not(HaveOccurred()))
-
-			checkKubeAPIServerDeployment(dep, annotations, "1.17.0", false)
-		})
-
 		It("should add missing elements to kube-apiserver deployment (k8s >= 1.18)", func() {
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer_test.go:121]:
```suggestion
		It("should add missing elements to kube-apiserver deployment (k8s = 1.20)", func() {
```

File: pkg/webhook/controlplane/ensurer_test.go [hunk 10]
```diff
@@ -312,38 +205,18 @@ var _ = Describe("Ensurer", func() {
...
-			client.EXPECT().Get(ctx, secretKey, &corev1.Secret{}).DoAndReturn(clientGet(secret))
-			client.EXPECT().Get(ctx, cmKey, &corev1.ConfigMap{}).DoAndReturn(clientGet(cm))
-
-			err := ensurer.EnsureKubeControllerManagerDeployment(ctx, eContextK8s118, dep, nil)
-			Expect(err).To(Not(HaveOccurred()))
-
-			checkKubeControllerManagerDeployment(dep, annotations, kubeControllerManagerLabels, "1.18.0", false)
-		})
-
 		It("should add missing elements to kube-controller-manager deployment (k8s >= 1.18 w/ CSI annotation)", func() {
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer_test.go:208]:
```suggestion
		It("should add missing elements to kube-controller-manager deployment (k8s >= 1.20 w/ CSI annotation)", func() {
```

File: pkg/controller/worker/machines_test.go [hunk 11]
```diff
@@ -655,65 +625,18 @@ var _ = Describe("Machines", func() {
...
-							Maximum:              worker.DistributeOverZones(1, maxPool2, 2),
-							MaxSurge:             worker.DistributePositiveIntOrPercent(1, maxSurgePool2, 2, maxPool2),
-							MaxUnavailable:       worker.DistributePositiveIntOrPercent(1, maxUnavailablePool2, 2, minPool2),
 							Labels:               utils.MergeStringMaps(labels, map[string]string{"topology.ebs.csi.aws.com/zone": zone2}),
 							MachineConfiguration: machineConfiguration,
 						},
 					}
 				})
 
 				It("should return machine deployments with AWS CSI Label (k8s>=1.18)", func() {
```
Comment (ialidzhikov) [inline pkg/controller/worker/machines_test.go:634]:
```suggestion
				It("should return machine deployments with AWS CSI Label (k8s=1.20)", func() {
```

File: pkg/controller/controlplane/valuesprovider_test.go [hunk 12]
```diff
@@ -145,24 +144,6 @@ var _ = Describe("ValuesProvider", func() {
...
-						Pods: &cidr,
-					},
-					Kubernetes: gardencorev1beta1.Kubernetes{
-						Version: "1.17.1",
-					},
-				},
-			},
-		}
-
 		clusterK8sAtLeast118 = &extensionscontroller.Cluster{
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider_test.go:147]: Do we still need it? Can we make it `clusterK8sAtLeast120` and set the k8s version to 1.20.x?

File: pkg/webhook/controlplane/ensurer_test.go [hunk 13]
```diff
@@ -419,32 +289,18 @@ var _ = Describe("Ensurer", func() {
 			ensurer = NewEnsurer(logger)
 		})
 
-		It("should not add anything to kube-scheduler deployment (k8s < 1.18)", func() {
-			err := ensurer.EnsureKubeSchedulerDeployment(ctx, eContextK8s117, dep, nil)
+		It("should add missing elements to kube-scheduler deployment (k8s >= 1.18)", func() {
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer_test.go:292]:
```suggestion
		It("should add missing elements to kube-scheduler deployment (k8s = 1.20)", func() {
```

File: pkg/webhook/controlplane/ensurer_test.go [hunk 14]
```diff
@@ -719,8 +567,7 @@ done
 				Expect(&kubeletConfig).To(Equal(newKubeletConfig))
 			},
 
-			Entry("kubelet < 1.18", eContextK8s117, semver.MustParse("1.17.0"), false, "", nil),
-			Entry("1.18 <= kubelet < 1.21", eContextK8s118, semver.MustParse("1.18.0"), true, "CSIMigrationAWSComplete", nil),
+			Entry("1.18 <= kubelet < 1.21", eContextK8s120, semver.MustParse("1.18.0"), true, "CSIMigrationAWSComplete", nil),
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer_test.go:570]: You could adapt to 1.20.

**Review by dimitar-kostadinov — COMMENTED** [review]
API refs to kubernetes docs here:
https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/hack/api-reference/api.json#L12

https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/hack/api-reference/config.json#L20
could be updated to 1.25

This check here: https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/pkg/webhook/controlplane/ensurer_test.go#L690 can be deleted

The mock client here: https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/pkg/webhook/controlplane/ensurer_test.go#L94
is not used and can be deleted, and the `InjectClient` function can be dropped as well.
https://github.com/gardener/gardener-extension-provider-aws/blob/c9a30abe43a0523a9b25ab85935e141cc7ba6cf4/pkg/webhook/controlplane/ensurer.go#L54

File: pkg/admission/validator/shoot.go (hunk 2 above)
Comment (MartinWeindel) [reply pkg/admission/validator/shoot.go:130]: Checked for our landscapes. The annotation `alpha.csimigration.shoot.extensions.gardener.cloud/kubernetes-version`  is not used for any shoot.

**Comment by MartinWeindel** [comment]
@ialidzhikov @dimitar-kostadinov 
Thanks for your reviews. I have addressed all your points.

**Review by dimitar-kostadinov — APPROVED** [review]
/lgtm

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**